### PR TITLE
Rustdoc: disambiguate Implementors when the type name is not unique

### LINKED ISSUE #37762: rustdoc should show module for types when the type name is not unique
If you look at the docs for [ExactSizeIterator](https://doc.rust-lang.org/stable/std/iter/trait.ExactSizeIterator.html) there are I think literally hundreds of implementations, with perhaps dozens that look identical, e.g. here are some lines I copied out:

```
impl<'a, T> ExactSizeIterator for Iter<'a, T>
impl<T> ExactSizeIterator for IntoIter<T>
impl<'a, T> ExactSizeIterator for Iter<'a, T>
impl<'a, T> ExactSizeIterator for IterMut<'a, T>
impl<T> ExactSizeIterator for IntoIter<T>
impl<T> ExactSizeIterator for IntoIter<T>
```

Those are all distinct implementations, but with the idiom of not-stuttering module names in type-names it is impossible to know which type you're looking for without hovering your mouse over each.

Those lines would be more helpful if they looked like:

```
impl<'a, T> ExactSizeIterator for btree_set::Iter<'a, T>
impl<T> ExactSizeIterator for btree_set::IntoIter<T>
impl<'a, T> ExactSizeIterator for linked_list::Iter<'a, T>
impl<'a, T> ExactSizeIterator for linked_list::IterMut<'a, T>
impl<T> ExactSizeIterator for linked_list::IntoIter<T>
impl<T> ExactSizeIterator for vec::IntoIter<T>
```

IMO the way that rustdoc should handle this is it should check if the type is ambiguous *at all* in the current documentation run (not just ambiguous on the current documentation page) and disambiguate with as many modules as necessary.

### PR DESCRIPTION
**Pull request by estebank**
Presentation [goes from](https://doc.rust-lang.org/stable/std/iter/trait.ExactSizeIterator.html#implementors):

to:

on cases where there're multiple implementors with the same name.

Fixes #37762.

### CODE REVIEW

**Comment by bluss** [comment]
Ignore-whitespace diff: https://github.com/rust-lang/rust/pull/38414/files?w=1

File: src/librustdoc/html/render.rs [hunk 1]
```diff
@@ -2111,9 +2111,23 @@ fn item_trait(w: &mut fmt::Formatter, cx: &Context, it: &clean::Item,
...
+            for l in implementors.iter() {
+                match (k.impl_.for_.clone(), l.impl_.for_.clone()) {
+                    (clean::Type::ResolvedPath {path: path_a, ..},
+                     clean::Type::ResolvedPath {path: path_b, ..}) => {
+                        if k.def_id != l.def_id && path_a.last_name() == path_b.last_name() {
+                            dissambiguate = true;
+                        }
+                    }
+                    _ => (),
+                }
```
Comment (bluss) [inline src/librustdoc/html/render.rs]: Is is quadratic (complexity) in the number of implementors, I don't think we can afford that. A trait like Clone only lists ~480 impls, but still.

Comment (bluss) [inline src/librustdoc/html/render.rs]: Unless I'm missing something, you can change `last_name()` to return `&str` and create a `BtreeMap<&str, usize>` or something to count everything's duplicitly here.

**Comment by bluss** [comment]
Can you apply this diff? Then it uses &str instead, and counts the right implementors. I wrote this up to make sure that it was possible (to not make suggestions that can't be).

```diff
diff --git a/src/librustdoc/clean/mod.rs b/src/librustdoc/clean/mod.rs
index 28ca92f..a99d4dd 100644
--- a/src/librustdoc/clean/mod.rs
+++ b/src/librustdoc/clean/mod.rs
@@ -2205,8 +2205,8 @@ impl Path {
         }
     }
 
-    pub fn last_name(&self) -> String {
-        self.segments.last().unwrap().name.clone()
+    pub fn last_name(&self) -> &str {
+        &self.segments.last().unwrap().name
     }
 }
 
diff --git a/src/librustdoc/html/render.rs b/src/librustdoc/html/render.rs
index 846c072..5839c46 100644
--- a/src/librustdoc/html/render.rs
+++ b/src/librustdoc/html/render.rs
@@ -2110,23 +2110,22 @@ fn item_trait(w: &mut fmt::Formatter, cx: &Context, it: &clean::Item,
         <h2 id='implementors'>Implementors</h2>
         <ul class='item-list' id='implementors-list'>
     ")?;
-    let mut implementor_count: FxHashMap<String, usize> = FxHashMap();
-    for (_, implementors) in cache.implementors.iter() {
+    if let Some(implementors) = cache.implementors.get(&it.def_id) {
+        let mut implementor_count: FxHashMap<&str, usize> = FxHashMap();
         for implementor in implementors {
             if let clean::Type::ResolvedPath {ref path, ..} = implementor.impl_.for_ {
                 *implementor_count.entry(path.last_name()).or_insert(0) += 1;
             }
         }
-    }
-    if let Some(implementors) = cache.implementors.get(&it.def_id) {
-        for implementor in implementors.iter() {
+
+        for implementor in implementors {
             write!(w, "<li><code>")?;
             // If there's already another implementor that has the same abbridged name, use the
             // full path, for example in `std::iter::ExactSizeIterator`
             let dissambiguate = if let clean::Type::ResolvedPath {
                 ref path, ..
             } = implementor.impl_.for_ {
-                *implementor_count.get(&path.last_name()).unwrap_or(&0) > 1
+                implementor_count[path.last_name()] > 1
             } else {
                 false
             };

```

**Comment by estebank** [comment]
@bluss applied 👍

File: src/librustdoc/html/format.rs [hunk 2]
```diff
@@ -433,7 +433,7 @@ pub fn href(did: DefId) -> Option<(String, ItemType, Vec<String>)> {
 /// Used when rendering a `ResolvedPath` structure. This invokes the `path`
 /// rendering function with the necessary arguments for linking to a local path.
 fn resolved_path(w: &mut fmt::Formatter, did: DefId, path: &clean::Path,
-                 print_all: bool) -> fmt::Result {
+                 print_all: bool, use_absolute: bool) -> fmt::Result {
```
Comment (bluss) [inline src/librustdoc/html/format.rs]: Can you go through the patch and use the same name for this parameter everywhere (the `use_absolute` here)? That's one step that makes a bool parameter more bearable.

File: src/librustdoc/html/format.rs [hunk 3]
```diff
@@ -551,194 +561,198 @@ impl<'a> fmt::Display for HRef<'a> {
...
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        match *self {
-            clean::Generic(ref name) => {
-                f.write_str(name)
-            }
-            clean::ResolvedPath{ did, ref typarams, ref path, is_generic } => {
-                // Paths like T::Output and Self::Output should be rendered with all segments
-                resolved_path(f, did, path, is_generic)?;
-                tybounds(f, typarams)
+fn fmt_type(t: &clean::Type, f: &mut fmt::Formatter, full_path: bool) -> fmt::Result {
```
Comment (bluss) [inline src/librustdoc/html/format.rs]: This parameter is the same (`full_path`), so this and in other places.

File: src/librustdoc/html/render.rs [hunk 4]
```diff
@@ -2111,9 +2111,25 @@ fn item_trait(w: &mut fmt::Formatter, cx: &Context, it: &clean::Item,
...
+                *implementor_count.entry(path.last_name()).or_insert(0) += 1;
+            }
+        }
+
+        for implementor in implementors {
             write!(w, "<li><code>")?;
-            fmt_impl_for_trait_page(&i.impl_, w)?;
+            // If there's already another implementor that has the same abbridged name, use the
+            // full path, for example in `std::iter::ExactSizeIterator`
+            let dissambiguate = if let clean::Type::ResolvedPath {
```
Comment (bluss) [inline src/librustdoc/html/render.rs]: Extra `s` in the spelling here. It's another of the ones that should have the same name across the patch, though.

**Review by bluss — CHANGES_REQUESTED** [review]
Thanks, we can integrate, let's just clean up that bool parameter's name to be consistent across the patch.

**Review by bluss — APPROVED** [review]

**Comment by steveklabnik** [comment]
@bors: r+

thanks!

**Comment by GuillaumeGomez** [comment]
Excellent, thanks a lot for this!

**Pull request merged**